documentation/testing.html

```diff
@@ -1,11 +1,69 @@
 ---
-# Copyright 2017 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
-title: "Unit Testing Configurable Components"
+# Copyright 2019 Oath Inc. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
+title: "Testing"
 ---
 
+
+<h2>Using Application </h2>
+<p>
+Notice the <code>com.yahoo.application.container.JDisc</code> class that is accessed by
+the test through <code>app.getJDisc(clusterName)</code>. That class has methods for
+exercising all common <a href="component-types.html">component types</a>.
+</p><p>
+Notice how the the test disables the network layer in the call to the
+<code>Application</code> factory method. Unit tests should disable
+the network to allow tests to be run in parallel. However, to have the JDisc
+Container bind to the appropriate network sockets, the network needs to be enabled.
+</p><p>
+Using Application instantiates the Container
+and lets you run requests using a browser or HTTP client,
+by enabling the network interface:
+<pre>
+public class ApplicationMain {
+
+  @Test
+  public static void main(String[] args) throws Exception {
+    try (com.yahoo.application.Application app = com.yahoo.application.Application.fromApplicationPackage(
+        FileSystems.getDefault().getPath("src/main/application"),
+        Networking.enable)) {
+      app.getClass(); // throws NullPointerException
+      Thread.sleep(Long.MAX_VALUE);
+    }
+  }
+}
+</pre>
+Run/debug this, and it will invoke the plugins as configured.
+Most often, you will write tests like:
+<pre>
+import java.io.File;
+import org.junit.Test;
+import com.yahoo.application.Application;
+import com.yahoo.search.Query;
+
+public class ApplicationTest {
+
+  @Test
+  public void testContainerApplicationWithReferencedContentCluster() throws Exception {
+    try (Application application =
+        Application.fromApplicationPackage(new File("myapplication/"),
+        Networking.disable) ) {
+      byte[] response = application.getJDisc("foo").search().processAndRender(chainId, rendererId,
```

Comment:
What is `foo` here?

Reply:
who knows. I did not touch this part (yet)

Reply:
just copied from elsewhere - will clean up